core/integration/bookkeeping/plugin.go

```diff
@@ -0,0 +1,319 @@
+	if p.bookkeepingClient == nil {
+		p.bookkeepingClient = Instance()
+		if p.bookkeepingClient == nil {
+			return fmt.Errorf("failed to connect to Bookkeeping service on %s", viper.GetString("bookkeepingBaseUri"))
+		}
+		log.Debug("Bookkeeping plugin initialized")
+	}
+	return nil
+}
+
+func (p *Plugin) ObjectStack(_ map[string]string) (stack map[string]interface{}) {
+	stack = make(map[string]interface{})
+	return stack
+}
+
+func (p *Plugin) CallStack(data interface{}) (stack map[string]interface{}) {
+	call, ok := data.(*callable.Call)
+	if !ok {
+		return
+	}
+	varStack := call.VarStack
+	envId, ok := varStack["environment_id"]
+	if !ok {
+		log.Error("cannot acquire environment ID")
+		return
+	}
+
+	stack = make(map[string]interface{})
+	stack["StartOfRun"] = func() (out string) { // must formally return string even when we return nothing
+		var err error
+		callFailedStr := "Bookkeeping StartOfRun call failed"
+
+		rn := varStack["run_number"]
+		var runNumber64 int64
+		runNumber64, err = strconv.ParseInt(rn, 10, 32)
+		if err != nil {
+			log.WithField("partition", envId).
+				WithError(err).
+				Error("cannot acquire run number for Bookkeeping SOR")
+		}
+
+		log.WithField("partition", envId).
+			WithField("level", infologger.IL_Ops).
+			WithField("runNumber", runNumber64).
+			Infof("performing Bookkeeping SOR")
+
+		if p.bookkeepingClient == nil {
+			err = fmt.Errorf("Bookkeeping plugin not initialized, StartOfRun impossible")
+
+			log.WithError(err).
+				WithField("level", infologger.IL_Support).
+				WithField("endpoint", viper.GetString("bookkeepingBaseUri")).
+				WithField("runNumber", runNumber64).
+				WithField("partition", envId).
+				WithField("call", "StartOfRun").
+				Error("Bookkeeping SOR error")
+
+			call.VarStack["__call_error_reason"] = err.Error()
+			call.VarStack["__call_error"] = callFailedStr
+
+			return
+		}
+
+		parsedEnvId, err := uid.FromString(envId)
+		envMan := environment.ManagerInstance()
+		env, err := envMan.Environment(parsedEnvId)
+		runNumber := env.GetCurrentRunNumber()
+
+		args := controlcommands.PropertyMap{
+			"runNumber": strconv.FormatUint(uint64(runNumber), 10),
+		}
+
+		flps := env.GetFLPs()
+		dd_enabled, _ := strconv.ParseBool(env.GetKV("", "dd_enabled"))
+		dcs_enabled, _ := strconv.ParseBool(env.GetKV("", "dcs_enabled"))
+		epn_enabled, _ := strconv.ParseBool(env.GetKV("", "epn_enabled"))
+		odc_topology := env.GetKV("", "odc_topology")
+		// GetString of active detectors and pass it to the BK API
+		detectors := strings.Join(env.GetActiveDetectors().StringList(), ",")
+
+		// TODO:
+		// add return error ???
+		p.bookkeepingClient.CreateRun(env.Id().String(), len(env.GetActiveDetectors()), 0, len(flps), int32(runNumber), env.GetRunType().String(), time.Now(), time.Now(), dd_enabled, dcs_enabled, epn_enabled, odc_topology, detectors)
+
+		for _, flp := range flps {
+			p.bookkeepingClient.CreateFlp(flp, flp, int32(runNumber))
```

Comment:
Let's ask George for a `CreateFlps` (plural) version of this. It's impractical to make 200 calls on every SOR.